RecoLocalTracker/Phase2TrackerRecHits/src/VectorHitBuilderAlgorithm.cc

```diff
@@ -0,0 +1,559 @@
+                                                                         const detset & theUpperDetSet)
+{
+  if(theLowerDetSet.size()==1 && theUpperDetSet.size()==1) return true;
+
+  //order lower clusters in u
+  std::vector<Phase2TrackerCluster1D> lowerClusters;
+  if(theLowerDetSet.size()>1) LogDebug("VectorHitBuilderAlgorithm") << " more than 1 lower cluster! " << std::endl;
+  if(theUpperDetSet.size()>1) LogDebug("VectorHitBuilderAlgorithm") << " more than 1 upper cluster! " << std::endl;
+  for ( const_iterator cil = theLowerDetSet.begin(); cil != theLowerDetSet.end(); ++ cil ) {
+    Phase2TrackerCluster1DRef clusterLower = edmNew::makeRefTo( clusters, cil );
+    lowerClusters.push_back(*clusterLower);
+  }
+  return true;
+}
+
+bool VectorHitBuilderAlgorithm::checkClustersCompatibility(Local3DPoint& poslower, 
+                                                           Local3DPoint& posupper, 
+                                                           LocalError& errlower, 
+                                                           LocalError& errupper)
+{
+  return true;
+
+}
+
+//----------------------------------------------------------------------------
+//ERICA::in the DT code the global position is used to compute the alpha angle and put a cut on that.
+std::vector<std::pair<VectorHit,bool>> VectorHitBuilderAlgorithm::buildVectorHits(const StackGeomDet * stack, 
+                                                                  edm::Handle< edmNew::DetSetVector<Phase2TrackerCluster1D> > clusters, 
+                                                                  const detset & theLowerDetSet, 
+                                                                  const detset & theUpperDetSet,
+                                                                  const std::vector<bool>& phase2OTClustersToSkip)
+{
+
+  std::vector<std::pair<VectorHit,bool>> result;
+  //ERICA: I am not sure why I introduced this first check. Maybe it is better to remove it..
+  if(checkClustersCompatibilityBeforeBuilding(clusters, theLowerDetSet, theUpperDetSet)){
+    LogDebug("VectorHitBuilderAlgorithm") << "  compatible -> continue ... " << std::endl;
+  } else { LogTrace("VectorHitBuilderAlgorithm") << "  not compatible, going to the next cluster"; }
+
+  std::vector<Phase2TrackerCluster1DRef> lowerClusters;
+  for ( const_iterator cil = theLowerDetSet.begin(); cil != theLowerDetSet.end(); ++ cil ) {
+    Phase2TrackerCluster1DRef clusterLower = edmNew::makeRefTo( clusters, cil );
+    lowerClusters.push_back(clusterLower);
+  }
+  std::vector<Phase2TrackerCluster1DRef> upperClusters;
+  for ( const_iterator ciu = theUpperDetSet.begin(); ciu != theUpperDetSet.end(); ++ ciu ) {
+    Phase2TrackerCluster1DRef clusterUpper = edmNew::makeRefTo( clusters, ciu );
+    upperClusters.push_back(clusterUpper);
+  }
+
+  std::sort(lowerClusters.begin(), lowerClusters.end(), LocalPositionSort(&*theTkGeom,&*cpe,&*stack->lowerDet()));
```

Comment:
Tracker geometry and cpe are already pointers, no need to dereference and take address again.